AppCenter/AppCenter/Internals/Context/AuthToken/MSAuthTokenContext.m

```diff
@@ -0,0 +1,70 @@
+#import "MSAuthTokenContext.h"
+#import "MSAuthTokenContextDelegate.h"
+
+/**
+ * Singleton.
+ */
+static MSAuthTokenContext *sharedInstance;
+static dispatch_once_t onceToken;
+
+@interface MSAuthTokenContext ()
+
+/**
+ * Collection of channel delegates.
+ */
+@property(nonatomic) NSHashTable<id<MSAuthTokenContextDelegate>> *delegates;
+
+@end
+
+@implementation MSAuthTokenContext
+
+@synthesize authToken = _authToken;
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _delegates = [NSHashTable new];
+  }
+  return self;
+}
+
++ (instancetype)sharedInstance {
+  dispatch_once(&onceToken, ^{
+    if (sharedInstance == nil) {
+      sharedInstance = [MSAuthTokenContext new];
+    }
+  });
+  return sharedInstance;
+}
+
++ (void)resetSharedInstance {
+  onceToken = 0;
+  sharedInstance = nil;
+}
+
+- (NSString *)authToken {
+  @synchronized(self) {
+    return _authToken;
+  }
+}
+
+- (void)setAuthToken:(NSString *)authToken {
```

Comment:
Is it okay to run this in main thread? Since browser is associated in login process, seems it will always be called in main thread. If we do flush during this time then it will block UI thread until it initiates flush.

Reply:
That is true. We can run on a background thread or expect `setAuthToken` to be invoked on a background thread. I prefer the latter because the caller has the responsibility to run this on a background thread, not the callee.

Reply:
:thinking: good point, my guess is it's probably multithreaded and synchronized but yes we should double check.